Dockerfile

```diff
@@ -51,14 +46,60 @@ ARG TAG
 WORKDIR /
 RUN git clone -b ${KUBERNETES_VERSION} --depth=1 https://github.com/kubernetes/kubernetes.git
 WORKDIR /kubernetes
-RUN make KUBE_GIT_VERSION=${TAG} GOLDFLAGS='-extldflags "-static"' GOFLAGS='-tags=netgo,osusergo' WHAT='cmd/kube-apiserver cmd/kube-controller-manager cmd/kube-proxy cmd/kube-scheduler cmd/kubeadm cmd/kubectl cmd/kubelet vendor/k8s.io/apiextensions-apiserver'
+RUN echo 'GO_BUILD_FLAGS=" \
+         -mod=vendor \
+         -gcflags=-trimpath=/kubernetes \
+         "' \
+    >> ./go-build-static
+RUN echo 'GO_LDFLAGS=" \
+         -X k8s.io/component-base/version.gitVersion=${TAG} \
+         -X k8s.io/component-base/version.gitCommit=$(git rev-parse HEAD) \
+         -X k8s.io/component-base/version.gitTreeState=clean \
+         -X k8s.io/component-base/version.buildDate=$(date -u +%Y-%m-%dT%H:%M:%SZ) \
+         -X k8s.io/client-go/pkg/version.gitVersion=${TAG} \
+         -X k8s.io/client-go/pkg/version.gitCommit=$(git rev-parse HEAD) \
+         -X k8s.io/client-go/pkg/version.gitTreeState=clean \
+         -X k8s.io/client-go/pkg/version.buildDate=$(date -u +%Y-%m-%dT%H:%M:%SZ) \
+         -linkmode=external -extldflags \"-static -Wl,--fatal-warnings\""' \
+    >> ./go-build-static
+RUN echo 'go build ${GO_BUILD_FLAGS} -ldflags "${GO_LDFLAGS}" "${@}"' \
+    >> ./go-build-static
+# force code generation
+RUN make WHAT=cmd/kube-apiserver
+# build statically linked executables
+RUN sh -ex ./go-build-static -o bin/kube-apiserver          ./cmd/kube-apiserver
+RUN sh -ex ./go-build-static -o bin/apiextensions-apiserver ./vendor/k8s.io/apiextensions-apiserver
+RUN sh -ex ./go-build-static -o bin/kube-controller-manager ./cmd/kube-controller-manager
+RUN sh -ex ./go-build-static -o bin/kube-proxy              ./cmd/kube-proxy
+RUN sh -ex ./go-build-static -o bin/kube-scheduler          ./cmd/kube-scheduler
+RUN sh -ex ./go-build-static -o bin/kubeadm                 ./cmd/kubeadm
+RUN sh -ex ./go-build-static -o bin/kubectl                 ./cmd/kubectl
+RUN sh -ex ./go-build-static -o bin/kubelet                 ./cmd/kubelet
+# assert statically linked executables
+RUN echo '[ -e $1 ] && (file $1 | grep -E "executable, x86-64, .*, statically linked")' \
```

Comment:
Does this catch everything? I had `file` report statically linked binaries while `nm` still showed some undefined and weak symbols.